### PR TITLE
Context to force a direct dial in the Swarm even if we have a Relayed connection

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For https://github.com/libp2p/go-libp2p/issues/1039.

### CODE REVIEW

**Review by vyzo — APPROVED** [review]

File: network/context.go [hunk 1]
```diff
@@ -12,8 +12,28 @@ var DialPeerTimeout = 60 * time.Second
 
 type noDialCtxKey struct{}
 type dialPeerTimeoutCtxKey struct{}
+type forceDirectDialCtxKey struct{}
```
Comment (marten-seemann) [inline network/context.go:15]: Why is the key a struct, not a `const string`?

Comment (marten-seemann) [reply network/context.go:15]: I see that it's consistent with what we've been doing so far, just wondering...

Comment (aschmahmann) [reply network/context.go:15]: The idea is you can't screw it up by accident. If you use a string someone else could use the string somewhere else and cause a conflict. For example, two applications could use the key word "forceDialDirect" one meaning to skip looking up the peer in some system like a DHT and the other meaning to refuse to use relays. Using an unexported struct makes this type of error basically impossible.

Comment (marten-seemann) [reply network/context.go:15]: That's clever. Thanks for the explanation!

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**